[LoadStoreVectorizer] Postprocess and merge equivalence classes

v-klochkov opened the pull request:
This patch introduces a new method:
void Vectorizer::mergeEquivalenceClasses(EquivalenceClassMap &EQClasses) const

The method is called at the end of Vectorizer::collectEquivalenceClasses() and is needed to merge equivalence classes that differ only by their underlying objects (UO1 and UO2), where UO1 is 1-level-indirection underlying base for UO2. This situation arises due to the limited lookup depth used during the search of underlying bases with llvm::getUnderlyingObject(ptr).

Using any fixed lookup depth can result into creation of multiple equivalence classes that only differ  by 1-level indirection bases.

The new approach merges equivalence classes if they have adjacent bases (1-level indirection). If a series of equivalence classes form ladder formed of 1-step/level indirections, they are all merged into a single equivalence class. This provides more opportunities for the load-store vectorizer to generate better vectors.

v-klochkov committed:
[LoadStoreVectorizer] Postprocess and merge equivalence classes

This patch introduces a new method:
void Vectorizer::mergeEquivalenceClasses(EquivalenceClassMap &EQClasses) const

The method is called at the end of Vectorizer::collectEquivalenceClasses() and is
needed to merge equivalence classes that differ only by their underlying objects
(UB1 and UB2), where UB1 is 1-level-indirection underlying base for UB2.
This situation arises due to the limited lookup depth used during the search of
underlying bases with llvm::getUnderlyingObject(ptr).

Using any fixed lookup depth can result into creation of multiple equivalence
classes that only differ  by 1-level indirection bases.

The new approach merges equivalence classes if they have adjucent bases (1-level indirection).
If a series of equivalence classes form ladder formed of 1-step/level indirections,
they are all merged into a single equivalence class.
This provides more opportunities for the load-store vectorizer to generate better vectors.

Signed-off-by: Klochkov, Vyacheslav N <vyacheslav.n.klochkov@intel.com>

arsenm commented on llvm/test/Transforms/LoadStoreVectorizer/X86/massive_indirection.ll:
```diff
@@ -0,0 +1,63 @@
+; common underlying object deeper than 6 levels from the original pointer.
+
+; The test below is the simplified version of actual performance oriented
+; workload; the offsets in getelementptr instructins are similar or same for
```
```suggestion
; workload; the offsets in getelementptr instructions are similar or same for
```

v-klochkov replied:
the misprint is fixed: https://github.com/llvm/llvm-project/pull/114501/commits/91a74c80e89ff46dcaf15dad30f9667e082d5979

arsenm commented on llvm/lib/Transforms/Vectorize/LoadStoreVectorizer.cpp:
```diff
@@ -1305,6 +1310,128 @@ std::optional<APInt> Vectorizer::getConstantOffsetSelects(
+
+  // For each item in RedKeyToUOMap, if it has more than one underlying object,
+  // try to merge the equivalence classes.
+  for (auto &RedKeyToUO : RedKeyToUOMap) {
+    auto UObjects = RedKeyToUO.second;
+    if (UObjects.size() < 2)
+      continue;
+    const auto RedKey = RedKeyToUO.first;
```
use c++17 structure binding

v-klochkov replied:
Thank you for the comment! I made an incremental fix for it: https://github.com/llvm/llvm-project/pull/114501/commits/91a74c80e89ff46dcaf15dad30f9667e082d5979

arsenm commented on llvm/lib/Transforms/Vectorize/LoadStoreVectorizer.cpp:
```diff
@@ -1305,6 +1310,128 @@ std::optional<APInt> Vectorizer::getConstantOffsetSelects(
+      continue;
+    const auto RedKey = RedKeyToUO.first;
+    auto UTMap = GetUltimateTargets(UObjects);
+    for (const auto &UT : UTMap) {
+      const Value *UObject = UT.first;
+      const Value *UltimateTarget = UT.second;
```
use c++17 structure binding

v-klochkov replied:
Fixed in https://github.com/llvm/llvm-project/pull/114501/commits/91a74c80e89ff46dcaf15dad30f9667e082d5979 . Thank you.

v-klochkov committed:
[NFC] Address review comments: fix misprint + use C++17 struct binding

arsenm commented on llvm/test/Transforms/LoadStoreVectorizer/X86/massive_indirection.ll:
```diff
@@ -0,0 +1,63 @@
+; RUN: opt %s -mtriple=x86_64-unknown-linux-gnu -passes=load-store-vectorizer -mcpu=skx -S -o %t.out.ll
+; RUN: FileCheck -input-file=%t.out.ll %s
```
Avoid the temporary file, nearly all tests should just pipe directly to FileCheck

v-klochkov replied:
Fixed as suggested: https://github.com/llvm/llvm-project/pull/114501/commits/427983a9378f8612c054ad219023564ecf803643

arsenm commented on llvm/test/Transforms/LoadStoreVectorizer/X86/massive_indirection.ll:
```diff
@@ -0,0 +1,63 @@
+; CHECK-LABEL: @v1_v2_v4_v1_to_v8_levels_6_7_8_8
+; CHECK: store <8 x half>
+
+  %level1 = getelementptr inbounds i8, ptr %arg1, i32 917504
+  %level2 = getelementptr i8, ptr %level1, i32 %arg0
+  %level3 = getelementptr i8, ptr %level2, i32 32768
+  %level4 = getelementptr inbounds i8, ptr %level3, i32 %arg0
+  %level5 = getelementptr i8, ptr %level4, i32 %arg0
```
Why do some of the GEPs have inbounds and some not? If they are not significant, they should probably be removed

v-klochkov replied:
There were some in the original workload. This test is the result of big simplification.
Those `inbound` are not important for the test, so I removed them as suggested: https://github.com/llvm/llvm-project/pull/114501/commits/427983a9378f8612c054ad219023564ecf803643

arsenm commented on llvm/test/Transforms/LoadStoreVectorizer/X86/massive_indirection.ll:
```diff
@@ -0,0 +1,63 @@
+
+define void @v1_v2_v4_v1_to_v8_levels_6_7_8_8(i32 %arg0, ptr align 16 %arg1) {
+; CHECK-LABEL: @v1_v2_v4_v1_to_v8_levels_6_7_8_8
+; CHECK: store <8 x half>
```
Use update_test_checks so we can see full context

v-klochkov replied:
Ok, added more context using `update_test_checks.py`. Thank you.

arsenm commented on llvm/lib/Transforms/Vectorize/LoadStoreVectorizer.cpp:
```diff
@@ -1305,6 +1310,124 @@ std::optional<APInt> Vectorizer::getConstantOffsetSelects(
+      EQClasses.erase(KeyFrom);
+    }
+  }
+  LLVM_DEBUG({
+    dbgs() << "LSV: mergeEquivalenceClasses: after merging:\n";
+    for (const auto &EC : EQClasses) {
+      dbgs() << "    Key: ([" << std::get<0>(EC.first)
+             << "]: " << *std::get<0>(EC.first) << ", " << std::get<1>(EC.first)
+             << ", " << std::get<2>(EC.first) << ", "
+             << static_cast<int>(std::get<3>(EC.first)) << ")\n";
+      for (const auto &Inst : EC.second)
+        dbgs() << "\tInst:\t" << *Inst << "\n";
+    }
+  });
```
```suggestion
  LLVM_DEBUG({
    dbgs() << "LSV: mergeEquivalenceClasses: after merging:\n";
    for (const auto &EC : EQClasses) {
      dbgs() << "    Key: ([" << std::get<0>(EC.first)
             << "]: " << *std::get<0>(EC.first) << ", " << std::get<1>(EC.first)
             << ", " << std::get<2>(EC.first) << ", "
             << static_cast<int>(std::get<3>(EC.first)) << ")\n";
      for (const auto &Inst : EC.second)
        dbgs() << "\tInst: " << *Inst << '\n';
    }
  });
```

arsenm commented on llvm/lib/Transforms/Vectorize/LoadStoreVectorizer.cpp:
```diff
@@ -1305,6 +1310,124 @@ std::optional<APInt> Vectorizer::getConstantOffsetSelects(
+                         std::get<2>(RedKey)};
+      EqClassKey KeyTo{UltimateTarget, std::get<0>(RedKey), std::get<1>(RedKey),
+                       std::get<2>(RedKey)};
+      auto VecFrom = EQClasses[KeyFrom];
+      auto VecTo = EQClasses[KeyTo];
```
No auto. Should these be references?

v-klochkov replied:
`const auto &` would be good here, IMO. Added it here: https://github.com/llvm/llvm-project/pull/114501/commits/427983a9378f8612c054ad219023564ecf803643

arsenm commented on llvm/lib/Transforms/Vectorize/LoadStoreVectorizer.cpp:
```diff
@@ -1305,6 +1310,124 @@ std::optional<APInt> Vectorizer::getConstantOffsetSelects(
+  if (!FoundPotentiallyOptimizableEC)
+    return;
+
+  LLVM_DEBUG({
+    dbgs() << "LSV: mergeEquivalenceClasses: before merging:\n";
+    for (const auto &EC : EQClasses) {
+      dbgs() << "    Key: ([" << std::get<0>(EC.first)
+             << "]: " << *std::get<0>(EC.first) << ", " << std::get<1>(EC.first)
+             << ", " << std::get<2>(EC.first) << ", "
+             << static_cast<int>(std::get<3>(EC.first)) << ")\n";
+      for (const auto &Inst : EC.second)
+        dbgs() << "\tInst:\t" << *Inst << "\n";
+    }
+  });
+  LLVM_DEBUG({
+    dbgs() << "LSV: mergeEquivalenceClasses: RedKeyToUOMap:\n";
+    for (const auto &RedKeyToUO : RedKeyToUOMap) {
+      dbgs() << "    Reduced key: (" << std::get<0>(RedKeyToUO.first) << ", "
+             << std::get<1>(RedKeyToUO.first) << ", "
+             << static_cast<int>(std::get<2>(RedKeyToUO.first)) << ") --> "
+             << RedKeyToUO.second.size() << " underlying objects:\n";
+      for (auto UObject : RedKeyToUO.second)
+        dbgs() << "    [" << UObject << "]: " << *UObject << "\n";
+    }
+  });
```
```suggestion
  LLVM_DEBUG({
    dbgs() << "LSV: mergeEquivalenceClasses: before merging:\n";
    for (const auto &EC : EQClasses) {
      dbgs() << "    Key: ([" << std::get<0>(EC.first)
             << "]: " << *std::get<0>(EC.first) << ", " << std::get<1>(EC.first)
             << ", " << std::get<2>(EC.first) << ", "
             << static_cast<int>(std::get<3>(EC.first)) << ")\n";
      for (const auto &Inst : EC.second)
        dbgs() << "\tInst: " << *Inst << '\n';
    }
  });
  LLVM_DEBUG({
    dbgs() << "LSV: mergeEquivalenceClasses: RedKeyToUOMap:\n";
    for (const auto &RedKeyToUO : RedKeyToUOMap) {
      dbgs() << "    Reduced key: (" << std::get<0>(RedKeyToUO.first) << ", "
             << std::get<1>(RedKeyToUO.first) << ", "
             << static_cast<int>(std::get<2>(RedKeyToUO.first)) << ") --> "
             << RedKeyToUO.second.size() << " underlying objects:\n";
      for (auto UObject : RedKeyToUO.second)
        dbgs() << "    [" << UObject << "]: " << *UObject << '\n';
    }
  });
```

arsenm commented on llvm/lib/Transforms/Vectorize/LoadStoreVectorizer.cpp:
```diff
@@ -324,6 +324,11 @@ class Vectorizer {
       Instruction *ChainElem, Instruction *ChainBegin,
       const DenseMap<Instruction *, APInt /*OffsetFromLeader*/> &ChainOffsets);
 
+  /// Merges the equivalence classes if they have uderlying objects that differ
```
```suggestion
  /// Merges the equivalence classes if they have underlying objects that differ
```

arsenm commented on llvm/lib/Transforms/Vectorize/LoadStoreVectorizer.cpp:
```diff
@@ -1305,6 +1310,124 @@ std::optional<APInt> Vectorizer::getConstantOffsetSelects(
+    for (const auto *UObject : UObjects) {
+      const unsigned MaxLookupDepth = 1; // look for 1-level indirections only
+      const auto *UltimateTarget =
+          llvm::getUnderlyingObject(UObject, MaxLookupDepth);
```
```suggestion
          getUnderlyingObject(UObject, MaxLookupDepth);
```

arsenm commented:
> Using any fixed lookup depth can result into creation of multiple equivalence classes that only differ by 1-level indirection bases.

The test IR doesn't look canonical / optimal. Other passes are expected to have rewritten the addressing expressions to be a simpler form. If I run your examples through -O3, it mostly vectorizes with some quirks:

1. have to hackily avoid a memset that forms
2. Ends up with a scalar store + a 7x vector store

v-klochkov committed:
[NFC] Minor fixes to address the 2nd round of review comments

v-klochkov commented:
> > Using any fixed lookup depth can result into creation of multiple equivalence classes that only differ by 1-level indirection bases.
> 
> The test IR doesn't look canonical / optimal. Other passes are expected to have rewritten the addressing expressions to be a simpler form. If I run your examples through -O3, it mostly vectorizes with some quirks:
> 
> 1. have to hackily avoid a memset that forms
> 2. Ends up with a scalar store + a 7x vector store

@arsenm : Matt - thank you for the review.
I added an incremental commit https://github.com/llvm/llvm-project/pull/114501/commits/427983a9378f8612c054ad219023564ecf803643 to address all of your comments.

Regarding the case itself. The original workload was way more complex, the LIT only shows the idea of it, it is a huge simplification.
For the 1st case/pattern I saw `7x` + `1x` vectors (for loads and stores), which ruined performance - `7x` vectors had to be lowered/legalized to `4x`+`2x`+`1x` producing 4 mem-ops + extra reg-to-reg moves and swizzles instead of a single `8x` mem-operation.

The 2nd pattern in LIT shows potentially worse situation, giving 8 equivalence classes having 1 scalar mem-operation in each.

michalpaszkowski reviewed (approved):
Thank you for the review @arsenm, please post any other comments/suggestions or let me know if anything wasn't addressed.

The PR looks good to me. @v-klochkov does not have write access, I can merge the PR early next week (waiting in case there are any other comments).

arsenm commented:
> Regarding the case itself. The original workload was way more complex, the LIT only shows the idea of it, it is a huge simplification. For the 1st case/pattern I saw `7x` + `1x` vectors (for loads and stores), which ruined performance - `7x` vectors had to be lowered/legalized to `4x`+`2x`+`1x` producing 4 mem-ops + extra reg-to-reg moves and swizzles instead of a single `8x` mem-operation.
> 

Can you add a case that's more representative? You can try llvm-reduce to get a sample

v-klochkov commented:
> > Regarding the case itself. The original workload was way more complex, the LIT only shows the idea of it, it is a huge simplification. For the 1st case/pattern I saw `7x` + `1x` vectors (for loads and stores), which ruined performance - `7x` vectors had to be lowered/legalized to `4x`+`2x`+`1x` producing 4 mem-ops + extra reg-to-reg moves and swizzles instead of a single `8x` mem-operation.
> 
> Can you add a case that's more representative? You can try llvm-reduce to get a sample

It may be challenging due to couple of reasons. There is a legal consideration: using llvm-reduce on non-disclosed code might not make it shareable. and second - the original input code is for a target that is not published yet, manually removing those target specifics details may result into pretty synthetic test again.

The existing LIT test is designed to be easily understandable. Adding more variations into the address-operation (use different offsets in getelementptr, or using slightly different operations - not only getelementptr), keeps the idea of the test the same.

Perhaps, we can both agree that the 6-steps-depth-limited search of the underlying object can easily lead to situations where equivalence classes have underlying objects that are only one step or level apart.

@arsenm - I am open to adding more variations in address operations if you believe it would improve the LIT test. Please let me know your thoughts first.

v-klochkov committed:
Add one more LIT test case

It is built from the original workload by copying the STORE instructions
to be optimized, and manually bulding missing IRs around them minimally
reproducing the context of the original STOREs in the full test.

v-klochkov commented:
> Can you add a case that's more representative? You can try llvm-reduce to get a sample

@arsenm - Matt - Please take a look at 1 extra LIT case here: https://github.com/llvm/llvm-project/pull/114501/commits/7b07f8cf1e9ad1345bcc97fe9c99d1bb7590d819

llvm-reduce is a great tool, I played with it; In this case though, it turned out to be easier to copy the interesting STOREs operations and minimally re-build missing IRs around them to show the idea of the context the STOREs in the original workload were placed into.
It is easy to get underlying objects that 1 indirection level apart from each other when the 1st store in a sequence is a store indexed by `index` and consequent stores indexed by `index` + `some_const`, which gives extra depth/level created by getelementptr `index + some_const`.

v-klochkov commented:
> > Can you add a case that's more representative? You can try llvm-reduce to get a sample
> 
> @arsenm - Matt - Please take a look at 1 extra LIT case here: [7b07f8c](https://github.com/llvm/llvm-project/commit/7b07f8cf1e9ad1345bcc97fe9c99d1bb7590d819)
> 
> llvm-reduce is a great tool, I played with it; In this case though, it turned out to be easier to copy the interesting STOREs operations and minimally re-build missing IRs around them to show the idea of the context the STOREs in the original workload were placed into. It is easy to get underlying objects that 1 indirection level apart from each other when the 1st store in a sequence is a store indexed by `index` and consequent stores indexed by `index` + `some_const`, which gives extra depth/level created by getelementptr `index + some_const`.

@arsenm - Greetings, this is just a reminder/ping message. Thank you!

michalpaszkowski reviewed (changes requested):
@v-klochkov Thank you for adding the test and addressing the comments! I just ran the code formatter. One more thing, could you please name the unnamed/numbered vregs in the test -- this should help with any changes in the future? Please rebase on top of `main` branch -- once the builbots pass, I will merge the change.

v-klochkov committed:
Fix LIT: Rename unnamed/numbered vregs in LIT; remove usage of undef

v-klochkov commented:
> I just ran the code formatter. One more thing, could you please name the unnamed/numbered vregs in the test -- this should help with any changes in the future? Please rebase on top of `main` branch -- once the builbots pass, I will merge the change.

@michalpaszkowski, thank you! I re-based the patch, removed unnamed/numbered vregs, removed `undef`: 4bff74970fc7d80dd5d261770d026c58b4b9a22f

michalpaszkowski reviewed (approved):
Thank you @v-klochkov! LGTM!

michalpaszkowski commented:
@v-klochkov Reverting the change due to [a failure](https://lab.llvm.org/buildbot/#/builders/55/builds/4171). This may require some investigation. Here is the [reverting pull request](https://github.com/llvm/llvm-project/pull/119657).

v-klochkov commented:
> @v-klochkov Reverting the change due to [a failure](https://lab.llvm.org/buildbot/#/builders/55/builds/4171). This may require some investigation. Here is the [reverting pull request](https://github.com/llvm/llvm-project/pull/119657).

@michalpaszkowski - Thank you for the quick revert of that patch.
I cherry-picked this patch and added extra fix for the memory error: https://github.com/llvm/llvm-project/pull/121861
Please see: https://github.com/llvm/llvm-project/pull/121861#issuecomment-257417626

The error is reproducible only with memory sanitizer. 
Is there a way to run one of post-commit CI bots before the commit there?

akshayrdeodhar commented:
@v-klochkov - I'm curious- just like for any fixed depth, we will be able to construct a pathological case where multiple equivalence classes will be created for the same base, couldn't we construct pathological cases for MergeEquivalenceClasses where the ladder of depth-1 jumps does not suffice either?

v-klochkov commented:
> @v-klochkov - I'm curious- just like for any fixed depth, we will be able to construct a pathological case where multiple equivalence classes will be created for the same base, couldn't we construct pathological cases for MergeEquivalenceClasses where the ladder of depth-1 jumps does not suffice either?

Not sure I understand the question. The patch resolved the issue that came from the fact of having limited/fixed depth of the underlying-object (UO) search in the beginning/creating the eq classes. If the found UOs differ by 1 level  (e.g. UO2 is GEP of UO1), then MergeEquivalenceClasses should merge them and increase chances for vectorization. If it the closest distance between two UOs is 2 steps (GEP-GEP), they will not be merged. IMO, that is possible, but unlikely scenario.

@akshayrdeodhar - do you have some test or workload that can be vertorized, but suffers from 2+ diff between originally found UOs?

nikic commented:
I'm somewhat confused by this change. Why add this kind of complex post-processing step instead of increasing the getUnderlyingObject() limits?

In fact, I did raise it from 6 to 10 recently (https://github.com/llvm/llvm-project/pull/143714). Is it possible that this functionality is now no longer needed?
